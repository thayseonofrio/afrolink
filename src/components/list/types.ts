
export enum ExperienceFilter {
    LessThanOne = "less-than-one",
    OneToTwo = "one-to-two",
    TwoToFour = "two-to-four",
    MoreThanFive = "more-than-five",
    MoreThanTen = "more-than-ten"
  }
