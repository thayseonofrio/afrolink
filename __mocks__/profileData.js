module.exports = {
	city: "Porto Alegre",
	country: "Brasil",
	experience: 12,
	gender: "Feminino",
	jobTitle: ["Desenvolvedora"],
	name: "Teste",
	skills: ["MongoDB", "JavaScript"],
	socialLinks: {email: "test@hotmail.com"},
	state: "RS"
}