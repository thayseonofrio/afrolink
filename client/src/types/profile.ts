type SocialLinksType = {
    [key: string]: string
}

export type ProfileType = {
    city: string,
    country: string,
    experience: number,
    gender: string,
    jobTitle: string[],
    name: string,
    skills: string[],
    socialLinks: SocialLinksType,
    state: string
}
