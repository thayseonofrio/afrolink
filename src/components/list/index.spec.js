import React from "react"
import apis from "../../services/profile"
import List from "."
import { fireEvent, render } from "@testing-library/react"
import profileData from "../../../__mocks__/profileData"
import {getFilteredProfiles} from "."
import { ExperienceFilter } from "./types"

jest.mock("../../services/profile")

describe("<List />", () => {
	const response = {data: [profileData]}
	beforeEach(() => {
		apis.getAllProfiles.mockImplementationOnce(() => Promise.resolve(response))
	})

	it("should render list with profile", async () => {
		const { container, findByText} = render(<List />)
		expect( await findByText("Desenvolvedora")).toBeInTheDocument()
		expect( await findByText("Feminino", { exact: false})).toBeInTheDocument()
		expect( await findByText("Teste")).toBeInTheDocument()
		expect(container).toMatchSnapshot()
	})

	it("should return filtered profiles by gender", async () => {
		const secondProfile = {
			...response.data,
			gender: "masculino"
		}
		const thirdProfile = {
			...secondProfile,
			name: "outra pessoa"
		}
		const initialProfiles = [...response.data, secondProfile, thirdProfile]
		const filteredProfiles = getFilteredProfiles(initialProfiles, "feminino")
		expect(filteredProfiles).toHaveLength(1)
		expect(filteredProfiles[0].gender).toBe("feminino")
	})

	it("should return filtered profiles by skills", async () => {
		const secondProfile = {
			...response.data,
			skills: ["JavaScript", "Python"]
		}
		const thirdProfile = {
			...secondProfile,
			skills: ["Java"]
		}
		const initialProfiles = [...response.data, secondProfile, thirdProfile]
		const filteredProfiles = getFilteredProfiles(initialProfiles, null, null, ["JavaScript"])
		expect(filteredProfiles).toHaveLength(2)
		expect(filteredProfiles[0].skills).toContain("JavaScript")
	})

	it("should return filtered profiles by state", async () => {
		const secondProfile = {
			...response.data,
			state: "RS"
		}
		const thirdProfile = {
			...secondProfile,
			state: "RJ"
		}
		const initialProfiles = [...response.data, secondProfile, thirdProfile]
		const filteredProfiles = getFilteredProfiles(initialProfiles, null, "RS")
		expect(filteredProfiles).toHaveLength(2)
		expect(filteredProfiles[0].state).toBe("RS")
	})

	it("should return filtered profiles by experience", async () => {
		const secondProfile = {
			...response.data,
			experience: 9
		}
		const thirdProfile = {
			...secondProfile,
			experience: 14
		}
		const initialProfiles = [...response.data, secondProfile, thirdProfile]
		const filteredProfiles = getFilteredProfiles(initialProfiles, null, null, null, ExperienceFilter.OneToTwo)
		expect(filteredProfiles).toHaveLength(2)
		expect(filteredProfiles[0].experience).toBe(12)
	})

	it("should return filtered profiles by state, gender, skills and experience", async () => {
		const secondProfile = {
			...response.data,
			gender: "masculino",
			state: "RS",
			skills: ["Java"],
			experience: 27
		}
		const thirdProfile = {
			...secondProfile,
			gender: "feminino",
			state: "RJ",
			skills: ["Python", "SQL"],
			experience: 1
		}
		const initialProfiles = [...response.data, secondProfile, thirdProfile]
		const filteredProfiles = getFilteredProfiles(initialProfiles, "masculino", "RS", ["Java"], ExperienceFilter.TwoToFour)
		expect(filteredProfiles).toHaveLength(1)
		expect(filteredProfiles[0].gender).toBe("masculino")
		expect(filteredProfiles[0].state).toBe("RS")
		expect(filteredProfiles[0].skills).toContain("Java")
		expect(filteredProfiles[0].experience).toBe(27)
	})

	it("should return empty if no profiles are given", async () => {
		const filteredProfiles = getFilteredProfiles()
		expect(filteredProfiles).toHaveLength(0)
	})

	it("should filter by selected gender option", async () => {
		const {getByRole, findByRole, findAllByText} = render(<List />)
		const selectGender = getByRole("button", { name: /gênero/i })
		fireEvent.keyDown(selectGender, { keyCode: 40 })
		const option = await findByRole("option", { name: /feminino/i })
		fireEvent.click(option)
		expect(await findAllByText(/gênero: feminino/i)).toHaveLength(1)
	})

	it("should filter by selected state option", async () => {
		const {getByRole, findByRole, findAllByText} = render(<List />)
		const selectState = getByRole("button", { name: /estado/i })
		fireEvent.keyDown(selectState, { keyCode: 40 })
		const option = await findByRole("option", { name: /Rio Grande do Sul/i })
		fireEvent.click(option)
		expect(await findAllByText(/RS/i)).toHaveLength(1)
	})

	it("should filter by selected experience option", async () => {
		const {getByRole, findByRole, findAllByText} = render(<List />)
		const selectExperience = getByRole("button", { name: /experiência/i })
		fireEvent.keyDown(selectExperience, { keyCode: 40 })
		const option = await findByRole("option", { name: /1 a 2 anos/i })
		fireEvent.click(option)
		expect(await findAllByText(/Tempo de experiência: 1 ano/i)).toHaveLength(1)
	})
})