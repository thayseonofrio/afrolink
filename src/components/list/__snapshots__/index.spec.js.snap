// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render list with profile 1`] = `
<div>
  <div
    class="makeStyles-filters-3"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-filtersInput-4"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="gender-filter-input-label gender-filter"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="gender-filter"
        role="button"
        tabindex="0"
      >
        Gênero
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="gender-filter"
        tabindex="-1"
        value="default"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-filtersInput-4"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="state-filter-input-label state-filter"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="state-filter"
        role="button"
        tabindex="0"
      >
        Estado
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="state-filter"
        tabindex="-1"
        value="default"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-filtersInput-4"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="skills-label skills-filter"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="skills-filter"
        role="button"
        tabindex="0"
      >
        Habilidades
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="skills-filter"
        tabindex="-1"
        value="Habilidades"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
      >
        <div
          class="MuiPaper-root makeStyles-paper-2 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="makeStyles-root-5"
          >
            <h2
              class="makeStyles-title-7"
            >
               
              Teste
               
            </h2>
            <div
              class="makeStyles-info-6"
            >
              <span>
                 
                Desenvolvedora
                 
              </span>
              <span>
                Gênero: 
                feminino
              </span>
              <span>
                Tempo de experiência: 
                1 ano 
                 
              </span>
              <span>
                Local: 
                Porto Alegre
                /
                RS
              </span>
              <div
                class="makeStyles-skills-8"
              >
                <span
                  class="label"
                >
                   
                  MongoDB
                   
                </span>
                <span
                  class="label"
                >
                   
                  JavaScript
                   
                </span>
              </div>
            </div>
            <hr
              class="makeStyles-line-9"
            />
            <div
              class="makeStyles-socialLinks-10"
            >
              <a
                class="social"
                href="mailto:test@hotmail.com"
                target="_blank"
              >
                <img
                  alt="email link"
                  src="test-file-stub"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
